abaco.go

```diff
@@ -86,138 +318,79 @@ func (device *AbacoDevice) ReadAllPackets() ([]*packets.Packet, error) {
-		device.cgroups = append(device.cgroups, g)
-		device.nchan += g.nchan
-	}
-	sort.Sort(ByGroup(device.cgroups))
-	device.firstchan = device.cgroups[0].firstchan
-
-	// Verify that no channel #s appear in 2 groups.
-	highest := -1
-	for i, g := range device.cgroups {
-		if g.firstchan <= highest {
-			return fmt.Errorf("Channel group %d=%v but previous range ended at %d", i, g, highest)
-		}
-		highest = g.firstchan + g.nchan - 1
-	}
-
-	// Use the first and last timestamp to compute sample rate.
-	if tsInit.T != 0 && tsFinal.T != 0 {
-		dt := float64(tsFinal.T-tsInit.T) / tsInit.Rate
-		// TODO: check for wrap of timestamp if < 48 bits
-		// TODO: what if ts.Rate changes between Init and Final?
-
-		// Careful: assume that any missed packets had same number of samples as
-		// the packets that we did see. Thus find the average samples per packet.
-		dserial := snFinal - snInit
-		if dserial > 0 {
-			avgSampPerPacket := float64(samplesInPackets) / float64(packetsRead)
-			device.sampleRate = float64(dserial) * avgSampPerPacket / dt
-			fmt.Printf("Sample rate %.6g /sec determined from %d packets:\n\tΔt=%f sec, Δserial=%d, and %f samp/packet\n", device.sampleRate,
-				packetsRead, dt, dserial, avgSampPerPacket)
+			allPackets = append(allPackets, p...)
+			packetsRead = len(allPackets)
 		}
 	}
-
-	device.unwrap = make([]*PhaseUnwrapper, device.nchan)
-	for i := range device.unwrap {
-		device.unwrap[i] = NewPhaseUnwrapper(abacoFractionBits, abacoBitsToDrop)
-	}
-	return nil
+	return allPackets, nil
 }
 
-// enumerateAbacoDevices returns a list of abaco device numbers that exist
-// in the devfs. If /dev/xdmaX_c2h_0 exists, then X is added to the list.
-// Does not handle cards with suffix other than c2h_0.
-func enumerateAbacoDevices() (devices []int, err error) {
-	for cnum := 0; cnum < maxAbacoCards; cnum++ {
+// enumerateAbacoRings returns a list of abaco ring buffer numbers that exist
+// in the devfs. If /dev/xdmaX_c2h_0_description exists, then X is added to the list.
+// Does not handle cards with suffix other than *_c2h_0.
+func enumerateAbacoRings() (rings []int, err error) {
+	for cnum := 0; cnum < maxAbacoRings; cnum++ {
 		name := fmt.Sprintf("xdma%d_c2h_0_description", cnum)
 		if region, err := shm.Open(name, os.O_RDONLY, 0600); err == nil {
 			region.Close()
-			devices = append(devices, cnum)
+			rings = append(rings, cnum)
 		}
 	}
-	return devices, nil
+	return rings, nil
 }
 
-// AbacoSource represents all Abaco devices that can potentially supply data.
+//------------------------------------------------------------------------------------------------
+
+// AbacoSource represents all AbacoRing ring buffers that can potentially supply data, as well
+// as all AbacoGroups that are discovered in the SampleData phase.
 type AbacoSource struct {
-	devices     map[int]*AbacoDevice
-	Ndevices    int
-	active      []*AbacoDevice
+	Nrings int // number of available ring buffers
+	arings map[int]*AbacoRing
```

Comment:
are they not guaranteed to be sequentially ordered? I guess if we had four and chose to read 1 and 3 then they would not be sequentially numbered.